### PR TITLE
Clear the CallbackStatus entry from the map in FlushlessEventProducerHandler

### PR DESCRIPTION
**Pull request by vmaheshw**
ISSUE:
Stuck partitions were reported in prod-lor1 mm.one twice in last 1 week. During  both the instances, either the bad Kafka broker was abruptly removed or there were lots of URP(Under replicated partitions).

Root-cause:
 
a.      The initial suspicion was Kafka failing to ACK() for some of the send(), resulting in stuck partitions.
b.      There were logs in the files which should not be present:   
            2021/07/04 13:34:58.325 ERROR [FlushlessEventProducerHandler] [kafka-producer-network-thread | DatastreamLiKafkaProducer] [brooklin-service] [] Internal error: checkpoints should progress in increasing order. Resolved checkpoint as 966284416 which is less than current checkpoint of 966284432
2021/07/04 13:34:58.325 ERROR [FlushlessEventProducerHandler] [kafka-producer-network-thread | DatastreamLiKafkaProducer] [brooklin-service] [] Internal error: checkpoints should progress in increasing order. Resolved checkpoint as 966284363 which is less than current checkpoint of 966284416
 
This log means that the ACK received is for an offset smaller than the in-memory checkpoint in callbackStatus. This either means out of order offsets coming from kafka or the inflight message set has out-of-order messages for some reason.
 
<ol start="3" type="a" style="margin-bottom: 0in; caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-family: Calibri; font-size: 14px; font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; margin-top: 0in;"><li class="MsoListParagraph" style="margin: 0in; font-size: 12pt; font-family: Calibri, sans-serif;">What happened is: We did not clear the callbackStatus object for the topic Partition that was<span style="font-family: &quot;Courier New&quot;;"><span class="Apple-converted-space"> </span>rewind to last checkpoint</span>. So, it maintained all the older in-memory states and started working on top of it.</li></ol><p class="MsoNormal" style="margin: 0in; font-size: 12pt; font-family: Calibri, sans-serif; caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none;"><span style="font-size: 11pt;"><o:p> </o:p></span></p>

Action | InFlight Message Set | ACK Queue | Checkpoint | Poll() offset | Notes
-- | -- | -- | -- | -- | --
  | {3, 5, 6, 7} | {4} | 3 | 8 | Poll() offset is the offset to be polled by consumer.
Send failure for 5 | {3, 5, 6, 7} | {4} | 3 | 3 | Seek to last checkpoint
Send 3, 4, 5, 6, 7 | {3, 5, 6, 7, 4} | {4} | 3 | 8 | Add offset to inflight Queue
Ack for 3 | {5, 6, 7, 4} | {3,4} -> {} | 5 | 8 | Move the checkpoint to 5, since all the ACKs are smaller than the first inflight message.
Send failure for 4 | {5, 6, 7, 4} | {} | 5 | 5 | Seek to last checkpoint
Send 5,6,7,8 | {5, 6, 7, 4, 8} | {} | 5 | 9 | Add send offset to inflight queue
Ack 7 | {5,6,4,8} | {7} | 5 | 9 | Add 7 to ACK queue, since 7 > 5(first inflight message)
Ack 6 | {5,4,8} | {6,7} | 5 | 9 | Add 6 to ACK queue, since 6 > 5 (first inflight message)
Ack 5 | {4,8} | {5, 6, 7} | 5 | 9 | Add 5 to ACK queue, since 5 > 4 (first inflight message)
  |   |   | [The checkpoint will remain stuck because 4 was never sent again] |   |  

<p class="MsoNormal" style="margin: 0in; font-size: 12pt; font-family: Calibri, sans-serif; caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none;"><span style="font-size: 11pt;"><o:p> </o:p></span></p><p class="MsoNormal" style="margin: 0in; font-size: 12pt; font-family: Calibri, sans-serif; caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none;"><span style="font-size: 11pt;"><o:p> </o:p></span></p><p class="MsoNormal" style="margin: 0in; font-size: 12pt; font-family: Calibri, sans-serif; caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none;">So, this stuck partition issue happens when there are 2  consecutive {send() failures,<span class="Apple-converted-space"> </span><span style="font-family: &quot;Courier New&quot;;">seekToLastCheckpoint</span>() } without the task thread restarting. This is the main Rootcause behind the stuck partitions seen during shutdown when there were brokers removals from Kafka side.</p><p class="MsoNormal" style="margin: 0in; font-size: 12pt; font-family: Calibri, sans-serif; caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none;"><o:p> </o:p></p>

Solution:
 
To fix this: The ordering of inflight message set is extremely important in this algorithm and because the older states were not cleared, it resulted in undesired transition states.
 
Option 1: We will clear the callbackStatus entry for the topic Partition which are rewind to older checkpoint().
Option 2: Different data-structure that can maintain the ordering as well as remove the duplicates for inflight message set. Priority Queue can take care of ordering, but will have to be extended to avoid addition of duplicates.
 
For more deterministic behavior, my recommendation is Option(1).

### CODE REVIEW

File: datastream-kafka-connector/src/test/java/com/linkedin/datastream/connectors/kafka/mirrormaker/TestKafkaMirrorMakerConnectorTask.java [hunk 1]
```diff
@@ -1276,30 +1276,30 @@ public void testInFlightMessageCount() throws Exception {
     KafkaMirrorMakerConnectorTestUtils.produceEvents(spicyTopic, 1, _kafkaCluster);
 
     // verify that in-flight message count for each topic is 1
```
Comment (jzakaryan) [inline datastream-kafka-connector/src/test/java/com/linkedin/datastream/connectors/kafka/mirrormaker/TestKafkaMirrorMakerConnectorTask.java:1278]: nit: outdated comment

File: datastream-server-api/src/main/java/com/linkedin/datastream/server/FlushlessEventProducerHandler.java [hunk 2]
```diff
@@ -55,6 +55,13 @@ public void clear() {
     _callbackStatusMap.clear();
   }
 
+  /**
+   * Clear the source-partition entry from the _callbackStatusMap
+   */
+  public void clear(String source, int partition) {
+    _callbackStatusMap.remove(new SourcePartition(source, partition));
```
Comment (jzakaryan) [inline datastream-server-api/src/main/java/com/linkedin/datastream/server/FlushlessEventProducerHandler.java:62]: Just for clarification. Will this result in the correct key getting removed from the map? Does `SourcePartition` override `getHashKey()`?

**Review by jzakaryan — APPROVED** [review]
LGTM. Nit for updating comment in the test.

**Review by somandal — DISMISSED** [review]

File: datastream-server-api/src/main/java/com/linkedin/datastream/server/FlushlessEventProducerHandler.java (hunk 2 above)
Comment (vmaheshw) [reply datastream-server-api/src/main/java/com/linkedin/datastream/server/FlushlessEventProducerHandler.java:62]: SourcePartition extends Pair underneath and uses the default Hash() function at Object level. SourcePartition was in use as Key in the HashMap for a long time.

**Review by somandal — APPROVED** [review]

**Review by shrinandthakkar — APPROVED** [review]
lgtm!

**Pull request merged**